main.cpp

```diff
@@ -219,7 +219,55 @@ struct TIB {
 // Make this global to ease debugging
 TIB tib;
 
+static bool blockUpper2GB() {
+	DEBUG_LOG("Blocking upper 2GB address space\n");
+
+	// 32-bit windows only reserves the lowest 2GB of memory for use by a process (https://www.tenouk.com/WinVirtualAddressSpace.html)
+	// Linux, on the other hand, will happily allow nearly the entire 4GB address space to be used.
+	// In order to prevent windows programs from being very confused as to why it's being handed
+	// addresses in "invalid" memory, let's map the upper 2GB of memory to ensure libc can't allocate
+	// anything there
+	std::ifstream procMap("/proc/self/maps");
+	std::string procLine;
+
+	// Reserve an absurd amount to avoid any possible reallocation
+	procLine.reserve(0x8000);
+
+	unsigned int lastMapEnd = 0;
+
+	const unsigned int FILL_MEMORY_ABOVE = 0x80000000; // 2GB
+
+	while (getline(procMap, procLine)) {
```

Comment:
ifstream is probably using a buffer internally, so it would make sense to start by switch this to reading the entire file using `read()` calls and then parsing out lines of that by hand.

Reply:
Done!